pkg/job/getmetricdata/batching.go

```diff
@@ -0,0 +1,97 @@
+package getmetricdata
+
+import (
+	"math"
+	"time"
+
+	"github.com/nerdswords/yet-another-cloudwatch-exporter/pkg/model"
+)
+
+type WindowCalculator interface {
+	Calculate(period time.Duration, length time.Duration, delay time.Duration) (time.Time, time.Time)
+}
+
+type iteratorFactory struct {
+	metricsPerQuery  int
+	windowCalculator WindowCalculator
+}
+
+func (b iteratorFactory) Build(data []*model.CloudwatchData, jobMetricLength, jobMetricDelay int64) BatchIterator {
+	if len(data) == 0 {
+		return nothingToIterate{}
+	}
+
+	params := data[0].GetMetricDataProcessingParams
+	params.Length = jobMetricLength
+	params.Delay = jobMetricDelay
+
+	return NewSimpleBatchIterator(b.windowCalculator, b.metricsPerQuery, data, params)
+}
+
+type nothingToIterate struct{}
+
+func (n nothingToIterate) Size() int {
+	return 0
+}
+
+func (n nothingToIterate) Next() ([]*model.CloudwatchData, time.Time, time.Time) {
+	return nil, time.Time{}, time.Time{}
+}
+
+func (n nothingToIterate) HasMore() bool {
+	return false
+}
+
+type simpleBatchingIterator struct {
+	size            int
+	currentBatch    int
+	startTime       time.Time
+	endTime         time.Time
+	data            []*model.CloudwatchData
+	entriesPerBatch int
+}
+
+func (s *simpleBatchingIterator) Size() int {
+	return s.size
+}
+
+func (s *simpleBatchingIterator) Next() ([]*model.CloudwatchData, time.Time, time.Time) {
+	// We are out of data return defaults
+	if s.currentBatch >= s.size {
+		return nil, time.Time{}, time.Time{}
+	}
+
+	startingIndex := s.currentBatch * s.entriesPerBatch
+	endingIndex := startingIndex + s.entriesPerBatch
+	if endingIndex > len(s.data) {
+		endingIndex = len(s.data)
+	}
+
+	// TODO are we technically doing this https://go.dev/wiki/SliceTricks#batching-with-minimal-allocation and if not
+	// would it change allocations to do this ahead of time?
+	result := s.data[startingIndex:endingIndex]
+	s.currentBatch++
+	return result, s.startTime, s.endTime
+}
+
+func (s *simpleBatchingIterator) HasMore() bool {
+	return s.currentBatch < s.size
+}
+
+// NewSimpleBatchIterator returns an iterator which slices the data in place based on the metricsPerQuery.
+func NewSimpleBatchIterator(windowCalculator WindowCalculator, metricsPerQuery int, data []*model.CloudwatchData, params *model.GetMetricDataProcessingParams) BatchIterator {
```

Comment:
Wondering if we can remove the dependency on `WindowCalculator`, e.g. move it to the processor, calculate `startTime` and `endTime` in the processor and pass those two fixed values to the iterator.

Reply:
I could move the calculator out to the processor but I would still want to have the iterator return `([]*model.CloudWatchData, *model.GetMetricDataProcessingParams)` so its clearly provides the params for the batch. Otherwise, the processor has to index the first element for the params to get them for start + endtime and that is kinda awkward. 

I think it's important iterator owns the params like this though and it's the basis of the follow up which can produce more "correct" batches https://github.com/kgeckhart/yet-another-cloudwatch-exporter/compare/kgeckhart/getmetricdata-iterator...kgeckhart/batching-variable-start-end-time?expand=1

Reply:
IIUC though the core of the logic is in the factory.Build. Even the varyingTimeIterator doesn't use `windowCalculator` or `startTime` and `endTime` after they're set at creation time.

I guess what I see here is that iterator doesn't care much of start and end, that's why I was thinking that bit could be isolated more.

Reply:
I agree, I think the start + endTime calc can be moved to `processor`. I'm only highlighting the `iterator` should be responsible for saying "here's the params which need to be used" so instead of,

```
Next() ([]*model.CloudwatchData, time.Time, time.Time)
Next() ([]*model.CloudwatchData)
```
we get

```
//  "here's the next batch and the params for it"
Next() ([]*model.CloudwatchData, *model.GetMetricDataProcessingParams)
```